README.rst
CNXML Library
=============

This is a python library for working with CNXML content, which includes the CNXML and COLLXML formats.

Install
-------

This library requires Python >=3.5.

Install using::

   pip install -e "."

Tests
-----

Use the tox to run the tests::
  
  tox

If tox is not installed, run the following::

  pip install tox
  
Edit/Validate files using `atom.io <https://atom.io>`_
------------------------------------------------------

Preparation
^^^^^^^^^^^

#. Install https://atom.io
#. Get the cnxml RNG Schema files

   #. Download the most recent version from https://github.com/Connexions/cnxml/releases (click the "zip" link)

      - It should be in your Download foler
      - Move it to ~/.neb/

   #. Unzip the file
   #. It should have created a folder named something like `cnxml-2.0.0` (with `/cnxml/xml/cnxml/schema/rng/0.7/cnxml-jing.rng` in it)
   #. Rename the folder to be something you are unlikely to accidentally delete (like `cnxml-validation`)

      - Remember the name for later when you are editing `~/.atom/config.cson`

Install
^^^^^^^

#. Start up Atom
#. Install the `linter-autocomplete-jing` package

#. Type <kbd>⌘</kbd>+<kbd>,</kbd> (for Mac) to open Settings (or click **Atom**, **Preferences...** in the menu bar)

   #. Click **Install** in the left-hand-side
   #. Enter `linter-autocomplete-jing` and click **Install**
   #. **Alternative:** run `apm install linter-autocomplete-jing` from the commandline

#. Edit `~/.atom/config.cson` by clicking **Atom**, **Config** in the menu bar and add the following lines (at the bottom of this document)
#. Restart Atom
#. Open an unzipped complete-zip. (I run `atom ~/Downloads/col1234_complete` **From a terminal**)
#. Verify by opening an `index.cnxml` file and typing in `<figure>` somewhere in the file. If it is a valid location then it should auto-add `id=""` for you


Changes to `~/.atom/config.cson`::

    "*":
      core:
        customFileTypes:
    
          # Add this to the bottom of the customFileTypes area.
          # Note: Indentation is important!
          "text.xml": [
            "index.cnxml"
          ]
    
    
      # And then this to the bottom of the file
      # 1. Make sure "linter-autocomplete-jing" only occurs once in this file!
      # 1. make sure it is indented by 2 spaces just like it is in this example.
    
      "linter-autocomplete-jing":
        displaySchemaWarnings: true
        rules: [
          {
            priority: 1
            test:
              pathRegex: ".cnxml$"
            outcome:
              schemaProps: [
                {
                  lang: "rng"
                  path: "~/.neb/cnxml-validation/cnxml/xml/cnxml/schema/rng/0.7/cnxml-jing.rng"
                }
              ]
          }
        ]

License
-------

This software is subject to the provisions of the GNU Affero General
Public License Version 3.0 (AGPL). See license.txt for details.
Copyright (c) 2016 Rice University


CHANGES.rst
.. Use the following to start a new version entry:

   |version|
   ----------------------

   - feature message

3.1.9
-----

- Add structured parsing of licenses (#56)
- Ensure that https is acceptable in license urls (#57)

3.1.6
-----

- Fix derived metadata validation error (#49)

3.1.5
-----

- Require md:language for collxml but not cnxml (#39)

3.1.4
-----

- Make module titles in collection optional (#35)

3.1.3
-----

- Support using document title once it is removed from metadata (#33)
- Remove some metadata unused by git storage (#32)

3.1.2
-----

- Update schema and add support for slug in metadata (#30)

3.1.1
-----

- Add support for new optional metadata (#28)

3.1.0
-----

- Relax required information when parsing metadata (#25)

3.0.1
-----

- Re-release of 3.0.0

3.0.0
-----

- Add tests for cnxml <md:derived-from ...> validation
- Fix failing derived-from cnxml validation tests
- Use a self-closing md:derived-from as the valid repr
- Move the metadata parsing from
  `Press <https://github.com/openstax/cnx-press>`_ to cnxml (#20)
- Add a CLI utility for extracting cnxml metadata to json (#22)

2.2.0
-----

- Extend validation API to accept multiple documents to validate

2.1.1
-----

- Tweak mdml schema to properly allow any metadata element under derived-from

2.1.0
-----

- Remove external pathlib dependency
- Drop support for Python <= 2.7
- Use versioneer for project versioning
- Correct license in setup.py
- Add instructions to validate with the atom.io editor

2.0.0
-----

- Add a commandline interface for collxml validation
- Rename main cli function to make room for another entrypoint
- Make the metadata section required in collxml
- Add collxml validation function

1.1.0
-----

- Use a namedtuple for validation error line data.

1.0.0
-----

- Initial release.


cnxml/parse.py
import re
from functools import partial

from lxml import etree


__all__ = (
    'NSMAP',
    'make_cnx_xpath',
    'parse_metadata',
)


# XML namespace mapping used by lxml.etree for parsing and element lookup
NSMAP = {
    "bib": "http://bibtexml.sf.net/",
    "c": "http://cnx.rice.edu/cnxml",
    "cnxorg": "http://cnx.rice.edu/system-info",
    "col": "http://cnx.rice.edu/collxml",
    "data": "http://www.w3.org/TR/html5/dom.html#custom-data-attribute",
    "datadev": "http://dev.w3.org/html5/spec/#custom",
    "dc": "http://purl.org/dc/elements/1.1/",
    "epub": "http://www.idpf.org/2007/ops",
    "lrmi": "http://lrmi.net/the-specification",
    "m": "http://www.w3.org/1998/Math/MathML",
    "md": "http://cnx.rice.edu/mdml",
    "mod": "http://cnx.rice.edu/#moduleIds",
    "qml": "http://cnx.rice.edu/qml/1.0",
    "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
}


DEFAULT_LICENSE_TEXT = 'Creative Commons Attribution License (ASSUMED)'
DEFAULT_LICENSE_URL = 'http://creativecommons.org/licenses/by/4.0/'


def _maybe(vals):
    """Grab the first value if it exists."""
    try:
        return vals[0]
    except IndexError:
        return None


def make_cnx_xpath(elm_tree):
    """Makes an xpath function that includes the CNX namespaces.

    :param elm_tree: the xml element to begin the xpath from
    :type elm_tree: an element-like object from :mod:`lxml.etree`

    """
    return partial(elm_tree.xpath, namespaces=NSMAP)


def _squash_to_text(elm, remove_namespaces=False):
    if elm is None:
        return None
    value = [elm.text or '']
    for child in elm.getchildren():
        value.append(etree.tostring(child).decode('utf-8').strip())
    if remove_namespaces:
        value = [re.sub(' xmlns:?[^=]*="[^"]*"', '', v) for v in value]
    value = ''.join(value)
    return value


def _parse_license(license_el):
    default_license = (DEFAULT_LICENSE_TEXT, DEFAULT_LICENSE_URL)
    if license_el is None:
        return default_license
    url = license_el.attrib.get('url', None)
    if url is None:
        return default_license
    url = url.strip()
    if len(url) == 0:
        return default_license

    license_info_map = {
        'by': {
            'text': 'Creative Commons Attribution License',
            'versions': ['1.0', '2.0', '3.0', '4.0']
        },
        'by-nd': {
            'text': 'Creative Commons Attribution-NoDerivs License',
            'versions': ['1.0', '2.0']
        },
        'by-nd-nc': {
            'text': 'Creative Commons Attribution-NoDerivs-NonCommercial' +
                    ' License',
            # TODO: by-nd-nc version 1.0 does not exist, remove when we
            #       know it is safe
            'versions': ['1.0', '2.0']
        },
        'by-sa': {
            'text': 'Creative Commons Attribution-ShareAlike License',
            'versions': ['1.0', '2.0']
        },
        'by-nc': {
            'text': 'Creative Commons Attribution-NonCommercial License',
            'versions': ['1.0', '2.0']
        },
        'by-nc-sa': {
            'text': 'Creative Commons Attribution-NonCommercial-ShareAlike' +
                    ' License',
            'versions': ['4.0']
        }

    }
    # Example urls:
    # (no locale) http://creativecommons.org/licenses/by/4.0/
    # (w/ locale) http://creativecommons.org/licenses/by/4.0/deed.(...)
    is_localized = '/deed.' in url
    type_and_version = (
        url[:url.rindex('/deed.')]  # ignore /deed.(...)
        if is_localized else
        url.rstrip('/')  # ignore trailing '/'
    ).split('/')[-2:]
    if len(type_and_version) != 2 or 'creativecommons.org' not in url:
        raise Exception(f'Invalid license url: "{url}"')
    typ, ver = type_and_version
    # Even if the license is localized, it should have a valid type and
    # version in the url
    if(typ not in license_info_map or
            ver not in license_info_map[typ]['versions']):
        raise Exception('Unknown license type or version: ' +
                        f'{{ url: "{url}", type: "{typ}", version: "{ver}" }}')
    if is_localized:
        # In this instance, use the text in the element. Hopefully this text
        # has been translated. We could store translated versions of license
        # text here; however, that could quickly become burdensome.
        text = license_el.text
        if text is None:
            raise Exception(f'Expected license text for "{url}"')
        text = text.strip()
        if len(text) == 0:
            raise Exception(f'Expected license text for "{url}"')
    else:
        text = license_info_map[typ]['text']
    return (text, url)


def parse_metadata(elm_tree):
    """Given an element-like object (:mod:`lxml.etree`)
    lookup the metadata and return the found elements

    :param elm_tree: the root xml element
    :type elm_tree: an element-like object from :mod:`lxml.etree`
    :returns: common metadata properties
    :rtype: dict

    """
    xpath = make_cnx_xpath(elm_tree)
    role_xpath = lambda xp: tuple(  # noqa: E731
        (_maybe(xpath(xp)) or "").split()
    )

    license_text, license_url = _parse_license(_maybe(xpath('//md:license')))
    props = {
        'id': _maybe(xpath('//md:content-id/text()')),
        'uuid': _maybe(xpath('//md:uuid/text()')),
        'canonical_book_uuid': _maybe(
            xpath('//md:canonical-book-uuid/text()')
        ),
        'version': _maybe(xpath('//md:version/text()')),
        'created': _maybe(xpath('//md:created/text()')),
        'revised': _maybe(xpath('//md:revised/text()')),
        'title':
            _maybe(xpath('//md:title/text()')) or xpath('//c:title/text()')[0],
        'slug': _maybe(xpath('//md:slug/text()')),
        'license_url': license_url,
        'license_text': license_text,
        'language': _maybe(xpath('//md:language/text()')),
        'authors': role_xpath('//md:role[@type="author"]/text()'),
        'maintainers': role_xpath('//md:role[@type="maintainer"]/text()'),
        'licensors': role_xpath('//md:role[@type="licensor"]/text()'),
        'keywords': tuple(xpath('//md:keywordlist/md:keyword/text()')),
        'subjects': tuple(xpath('//md:subjectlist/md:subject/text()')),
        'abstract': _squash_to_text(
            _maybe(xpath('//md:abstract')),
            remove_namespaces=True,
        ),
        'print_style': _maybe(
            xpath('//col:param[@name="print-style"]/@value'),
        ),
        'derived_from': {
            'uri': _maybe(xpath('//md:derived-from/@url')),
            'title': _maybe(xpath('//md:derived-from/md:title/text()')),
        },
    }
    return props
